File: packages/javascript/src/store/store.test.ts
import Memory from './memory'

[
    Memory
]
    .map(mod => {
        describe(mod.name, () => {
            it('implementation', () => {
                const store = new mod()
                store.init()

                store.mutation({
                    merge: {
                        animals: {
                            shark: "hammerhead"
                        }
                    },
                    delete: {}
                })
                expect(store.query({})).toEqual({ animals: { shark: "hammerhead" } })

                store.mutation({
                    delete: {
                        animals: {
                            shark: 1
                        }
                    },
                    merge: {}
                })
                expect(store.query({})).toEqual({ animals: {} })

                store.mutation({
                    delete: { animals: {} },
                    merge: { animals: { fish: "barracuda", shark: "hammerhead" } }
                })
                expect(store.query({})).toEqual({ animals: { fish: "barracuda", shark: "hammerhead" } })
                expect(store.query({ animals: { shark: {} } })).toEqual({ animals: { shark: "hammerhead" } })
                expect(store.query({ animals: { stingray: { climate: {} } } })).toEqual({ animals: { stingray: { climate: undefined } } })

                expect(store.query_path(['animals', 'shark'])).toEqual('hammerhead')
                expect(store.query_keys(['animals'])).toEqual(['fish', 'shark'])
                store.merge(['animals', 'whale'], 'orca')
                expect(store.query_path(['animals', 'whale'])).toEqual('orca')
                store.delete(['animals', 'whale'])
                expect(store.query_path(['animals', 'whale'])).toEqual(undefined)
            })

        })

    })

File: packages/elixir/test/store_test.exs
defmodule Riptide.Test.Store do
  use ExUnit.Case

  test Riptide.Store.Memory, do: test_store(Riptide.Store.Memory, %{})

  defp test_store(store, opts) do
    :ok = store.init(opts)

    :ok =
      Riptide.Store.mutation(
        Riptide.Mutation.merge(["animals"], %{
          "shark" => "hammerhead",
          "whale" => "orca"
        }),
        store,
        opts
      )

    %{"animals" => %{"shark" => "hammerhead"}} =
      Riptide.Store.query(%{"animals" => %{"shark" => %{}}}, store, opts)

    2 = Riptide.Store.stream(["animals"]) |> Enum.count()
    1 = Riptide.Store.stream(["animals"], %{limit: 1}) |> Enum.count()
    [{"whale", _}] = Riptide.Store.stream(["animals"], %{min: "whale"}) |> Enum.to_list()

    %{
      "animals" => %{
        "shark" => "hammerhead",
        "whale" => "orca"
      }
    } =
      Riptide.Store.query(%{
        "animals" => %{
          "shark" => %{},
          "whale" => %{}
        }
      })
  end
end
